### ISSUE TITLE
Unable to reproduce the performance on ConSep Dataset

### ISSUE THREAD

**Issue by PrashastiSachan** [vqdang/hover_net#217]
Hi,

Thanks for sharing your work. I am using pytorch branch to train the model. However, it is not performing well on validation data during training.
The settings are same as your code in hover/opt.py and config.py
Am I missing any step to run before inference?
Here is the training metrics when training with the pretrained "npz" file
![image](https://user-images.githubusercontent.com/35075466/181780980-6a63e8e5-4950-4b34-811c-8d1d2f458b65.png)
![image](https://user-images.githubusercontent.com/35075466/181781066-399fd20a-33be-4aba-bf0d-ce04ce559ea2.png)
![image](https://user-images.githubusercontent.com/35075466/181781160-5efc948f-ff7e-471d-8667-4db6b8922588.png)
![image](https://user-images.githubusercontent.com/35075466/181781245-250cf02d-8c2c-4caf-bd81-73d1f690d6c2.png)
![image](https://user-images.githubusercontent.com/35075466/181781305-4c0c7555-e4a9-4e73-9a99-5d2eb524e6ac.png)
![image](https://user-images.githubusercontent.com/35075466/181781378-8465756d-e1fd-4e35-a827-3c59ebf66803.png)
![image](https://user-images.githubusercontent.com/35075466/181781730-4a162315-3771-4417-8941-5eb5f2f01d7e.png)
![image](https://user-images.githubusercontent.com/35075466/181781777-df004ad5-c670-4fd6-9d82-2989b009a305.png)

 Let me know how can I correct it to reproduce your results

**Comment by VolodymyrChapman** [comment]
I may have answered this in another issue: [issue 220](https://github.com/vqdang/hover_net/issues/220#issuecomment-1255921749)
Please confirm that you have used exactly the same learning process as stated in the original paper. By default, this HoVerNet GitHub has somewhat different parameters.

**Comment by zhouyi-git** [comment]
Why can't I find the value of the three output indicators of the test, for example, Dice=0.85, where is the output?

**Comment by hjtalent2023** [comment]
Hello! I have met a similarly question with you. like flow:
------valid-np_acc    : 0.92795
------valid-np_dice   : 0.73184
------valid-tp_dice_0 : 0.94518
------valid-tp_dice_1 : 0.02475
------valid-tp_dice_2 : 0.49475
------valid-tp_dice_3 : 0.00000
------valid-tp_dice_4 : 0.61169
------valid-hv_mse    : 0.06671
the dice_3 is always 0. I use the Consep dataset,and set the nr_types are 5. Does anybody can help me?

**Comment by kountaydwivedi** [comment]
@VolodymyrChapman 
Dear Sir

I am facing the same issue. I cloned the Hovernet github repo and did exactly as instructed -- setup the path to the directories and pretrained weights (ResNet) in config.py and setup the hyperparameters in the opt.py. I trained the training set of Consep dataset and then utilized net_epoch=50.tar checkpoint to infer the model on test set of Consep, but I am getting absolutely bizzare results. However, when I am using the provided consep checkpoint (hovernet_original_consep_type_tf2pytorch.tar), I am getting the results as presented in the paper.

I am attaching two images -- one is the output when inferred on net_epoch=50.tar (top) and the other when trained on the provided ckpt (hovernet_original_consep_type_tf2pytorch.tar) (bottom).

![Image](https://github.com/user-attachments/assets/51604192-f1a4-4b9c-95b3-69f23bc911ce)

![Image](https://github.com/user-attachments/assets/8d034fd3-6ebb-448d-b38b-65696ba02986)

**Comment by VolodymyrChapman** [comment]
Hi @kountaydwivedi , I hope you're well :-)
**Just to clarify for me, initial checkpoints for the images above are:**
- Top image: pretrained ImageNet weights for the ResNet?
- Bottom image: CoNSeP segmentation & classification checkpoint?

**With reference to the published ([HoVerNet training strategy](https://www.sciencedirect.com/science/article/pii/S1361841519301045?via=ihub#sec0005)):**
> Regarding HoVer-Net, we initialised the model with **pre-trained weights on the ImageNet dataset** ([Deng et al., 2009](https://www.sciencedirect.com/science/article/pii/S1361841519301045?via=ihub#bib0012)), trained **only the decoders** for the **first 50 epochs**, and then **fine-tuned all layers for another 50 epochs**........ Specifically, we used a batch size of 8 and 4 on each GPU for stage one and two respectively. We used Adam optimisation with an initial [learning rate](https://www.sciencedirect.com/topics/computer-science/learning-rate) of **10-4** and then reduced it to a rate of **10-5** after 25 epochs

**The following differences may therefore exist between your training approach and the pretrained checkpoint:**
- Initial checkpoint - is it ImageNet?
- training stage 1 for 50 epochs: are all parameters, other than the decoders, frozen for this stage? (param.requires_grad = False for rest) + batch size of 8/GPU?
- training stage 2 for 50 epochs: are all parameters 'unfrozen'? (param.requires_grad = True for all) + batch size of 4/GPU?

**Side notes about training continuity but please, anyone, chime in and tell me if I'm mistaken:** 
- The original paper used Nvidia 1080Ti GPUs (in TensorFlow; I'm making the assumption that similar GPUs were used for the pretrained PyTorch checkpoints). In my non-exhaustive experience, I have found Ampere GPUs (Nvidia 3000 series) onwards to give fairly different training results. May not apply in your case and may just be my situation but something to consider
- All dependency versions are pinned in the environment.yml and requirements.txt files so I don't think version differences are responsible for the training differences

I hope something in this long note helps :-)
My best wishes, 
Volodymyr

**Comment by kountaydwivedi** [comment]
Dear Sir

### Thank you for your prompt response. Regarding top image:

- Initial checkpoint for training: pretrained ResNet
- All other params/hyperparams: default (left as they are in the repository)
For the training purpose, I am just inputting the raw CoNSeP dataset path and the log path, and then executing the command `python run_train.py --gpu='0,1'`. Next, when the training is complete, I am utilizing the `np_epoch=50.tar` in the `01` folder in the logs directory to infer the model on the test dataset by executing the following script:

`python run_infer.py --gpu='0,1' --nr_types=5 --type_info_path=type_info.json --batch_size=64 --model_mode=original --model_path=original_hovernet_implementation/logs/01/net_epoch=50.tar --nr_inference_workers=8 --nr_post_proc_workers=16 tile --input_dir=original_hovernet_implementation/consep_dataset/Test/Images --output_dir=original_hovernet_implementation/consep_dataset/test_output --draw_dot --save_qupath`

### Now, regarding the bottom image:

All I did was to run the infer script on the test dataset:
`python run_infer.py --gpu='0,1' --nr_types=5 --type_info_path=type_info.json --batch_size=64 --model_mode=original --model_path=original_hovernet_implementation/hovernet_original_consep_type_tf2pytorch.tar --nr_inference_workers=8 --nr_post_proc_workers=16 tile --input_dir=original_hovernet_implementation/consep_dataset/Test/Images --output_dir=original_hovernet_implementation/consep_dataset/test_output/with_their_provided_ckpt --draw_dot --save_qupath`



I just want to point out one more thing-- I have not employed the authors' provided `extract_patches.py` script to extract patches. Rather, I created the `.npy` files using one of my own scripts on the raw images and annotations as follows:

![Image](https://github.com/user-attachments/assets/2c9ebe8c-e0ee-4cca-b380-841a4f028dfa)

I am sorry for pasting the code image; it seems that the code block was breaking a lot when pasting the code directly.

Thank you Sir for your help.
Sincerely
Kountay Dwivedi

**Comment by VolodymyrChapman** [comment]
Hi @kountaydwivedi , 
Thank you for clarifying. 
The key question is whether the model is training? i.e. whether performance visibly improves over epochs? 
**One way to do this is for you to:**
1. Train the pretrained ResNet model for X epochs (say, 25) 
2. Perform inference with the trained model to inspect performance
3. Train for another X epochs, starting from your new model (by modifying this path in opt.py: https://github.com/vqdang/hover_net/blob/67e2ce5e3f1a64a2ece77ad1c24233653a9e0901/models/hovernet/opt.py#L55)
4. Perform inference with the new model to inspect changes in model performance
5. Repeat steps 3 & 4 a few times to determine whether training is taking place

My best wishes,
Volodymyr
